ros/api/v1/recommendation_ratings.py

```diff
@@ -5,46 +5,90 @@
 from ros.lib.utils import identity, user_data_from_identity
 
 
-class RecommendationRatingsApi(Resource):
+def validate_rating_post_api(func):
+    """Validate POST rating request."""
+    allowed_choices = [c.value for c in RatingChoicesEnum]
 
-    rating_fields = {
-        'inventory_id': fields.String,
-        'rating': fields.Integer
-    }
+    def error_msg(error_code, value):
+        errors = {
+            400: "is invalid value for rating.",
+            422: "is invalid choice of input for rating."
+        }
+        return (
+            f"'{value}' { errors.get(error_code, 'Invalid') }"
+            f"Possible values - { *allowed_choices, }"
+        )
 
-    @marshal_with(rating_fields)
-    def post(self):
-        """
-            Add or update a rating for a system, by system ID.
-            Return the new rating. Any previous rating for this rule by this
-            user is amended to the current value.
-        """
+    def check_for_rating(data):
+        rating = None
+        try:
+            rating = int(data['rating'])
+        except ValueError:
+            abort(400, message=(error_msg(400, data['rating'])))
+
+        if rating not in allowed_choices:
+            abort(422, message=(error_msg(422, data['rating'])))
+
+        return rating
+
+    def check_for_user():
         ident = identity(request)['identity']
         user = user_data_from_identity(ident)
         username = user['username'] if 'username' in user else None
 
         if username is None:
             abort(403, message="Username doesn't exist")
 
-        data = json.loads(request.data)
-        inventory_id = data['inventory_id']
-        rating = int(data['rating'])
-        if rating not in [c.value for c in RatingChoicesEnum]:
-            abort(
-                422,
-                message=(
-                    "{} is not a valid value for rating".format(rating)))
+        return username
 
+    def check_for_system(inventory_id):
         system = System.query.filter(
             System.inventory_id == inventory_id
         ).first()
 
         if system is None:
-            abort(404, message="System {} doesn't exist"
-                  .format(inventory_id))
+            abort(404, message=f"System {inventory_id} doesn't exist")
+
+        return system.id
+
+    def validate_request(*args, **kwargs):
+        username = check_for_user()
+        data = json.loads(request.data)
```

Comment:
Tested w/o `inventory_id` and `rating`, getting a 500 server error on this line.

**Error:** json.decoder.JSONDecodeError: Expecting value

Happy path tests went fine :+1:

Reply:
Handle these exceptions too. Please request to review and confirm the changes.